### ISSUE TITLE
Support for `getServerSideProps` 

### ISSUE THREAD

**Issue by thomb** [isaachinman/next-i18next#652]
## Is your feature request related to a problem? Please describe.
When using `getServerSideProps` as opposed to `getInitialProps` results in a `namespacesRequired` warning, despite the fact that I am returning `namespacesRequired` inside the required `props` key

I don't know if this is factual or not, or if the warning is a red herring

## Describe the solution you'd like
Ideally I'd like `getServerSideProps` to be supported.

Apologies if this is vague or unclear, or a completely unreasonable request.

Thanks in advance

**Comment by isaachinman** [comment]
NextJs 9.3 was released a week ago. None of the new data fetching methods are supported yet, but in general I'd lean towards `getStaticProps` over `getServerSideProps`. Typically it's better to associate translation changes with releases, rather than having changes at runtime.

The `next-i18next` package will be entirely rewritten once NextJs plugin support lands, as quite a few things have changed since its inception.

**Comment by HsnVahedi** [comment]
I'm kind of newbie so sorry for asking this question. As I've figured out your amazing library currently fully supports Next js 9.2, right?

**Comment by isaachinman** [comment]
It fully supports NextJs 9.3 as well, just not the new data fetching methods.

**Comment by HsnVahedi** [comment]
In NextJs 9.3, Page components can not have getInitialProps. How could it support 9.3 ?

**Comment by isaachinman** [comment]
That is entirely false. That'd be a breaking change, and would go against semver. Read the docs.

> No deprecations are introduced and getInitialProps will continue to function as it currently does.

**Comment by kahoowkh** [comment]
> It fully supports NextJs 9.3 as well, just not the new data fetching methods.

May I know if there is any schedule for supporting `getServerSideProps` and `getStaticProps`?

**Comment by isaachinman** [comment]
I do not currently have much time for development on this project due to other factors in life. I don't see this changing for the next few weeks, at least.

As I've stated, `next-i18next` will be rewritten once plugin support (finally) lands. If you'd like to see this (and many other features) supported, please let the NextJs team know that you are waiting for plugin support!

**Comment by maxmalov** [comment]
Just noticed that you can use `i18next` inside `getServerSideProps` in the same way as you would do it in the express application, i.e. create & configure plain `i18next` instance and use it instead

**Comment by isaachinman** [comment]
@maxmalov Can you explain what you mean? We still have a dependency on the i18next browser detector middleware.

**Comment by maxmalov** [comment]
`getServerSideProps` is being called on the server-side only, consider the example

```js
import i18next from 'i18next';

const createI18Next = (lang) => {
  // create & configure i18next instance 
}

export const serverSideMethod = ({ params, req, res }) => {
  // detect lang from req somehow (probably can reuse express middleware)
  const lang = detect(req);
  const i18n = createI18Next(lang);
  const { t } = i18n;

  // do whatever you want with t
} 
```

So later one can use `serverSideMethod` inside `getServerSideProps` and get already translated content. For now, **it's just a theory**. We're working on the AMP first app and have to reuse the existing 18next resources we use in the separate single-page app. So in a few days, we eventually find out whether it works or not

**Comment by isaachinman** [comment]
Yes, that approach will work if you want to roll your own implementation, but has nothing to do with this package. Let's keep discussion here on topic.

**Comment by pettomartino** [comment]
Any updates here? 
It can be a workaround for now, or a way to silence

> You have not declared a namespacesRequired array on your page-level component: Page. This will cause all namespaces to be sent down to the client, possibly negatively impacting the performance of your app. For more info, see: https://github.com/isaachinman/next-i18next#4-declaring-namespace-dependencies"

Thanks

**Comment by isaachinman** [comment]
Hey @pettomartino - no update from my side. You can always silence warnings by setting `strictMode` to `false`. Please do feel free to contribute if you so choose!

**Comment by pettomartino** [comment]
@isaachinman thank for so short reply. Gonna try to implement something then. Any hints?

**Comment by isaachinman** [comment]
@pettomartino Do you mean contributing to `next-i18next`? If we want to support `getServerSideProps`, we'll need to update the `appWithTranslation` HOC to also look for that data fetching method in addition to `getInitialProps`.

**Comment by s875515** [comment]
There is another library solution, but it needs you to change the library...
https://github.com/zeit/next.js/blob/canary/examples/with-i18n-rosetta/README.md

**Comment by s875515** [comment]
@pettomartino there is a demonstration that the i18n solution works in SSG. Hope it helps
https://github.com/zeit/next.js/pull/11841

**Comment by icrona** [comment]
since we cant use serverSideProps together with initialProps. we can make use of React FunctionComponent.defaultProps
here is the workaround that works for me, and it still works with initialProps (https://github.com/isaachinman/next-i18next/issues/615#issuecomment-575578375)

1. particular page
`Page.defaultProps = {
      i18nNamspacves: ['home-page']
}`

2. main app / _app.tsx
`MyApp.getInitialProps = async (appContext: AppContextType<Router>) => {
          const appProps = await App.getInitialProps(appContext)

          const defaultProps = appContext.Component.defaultProps

      return {
        ...appProps,
        pageProps: {
            namespacesRequired: [...(appProps.pageProps.namespacesRequired || []), ...(defaultProps?.i18nNamespaces || [])],
}`

**Comment by icecold21** [comment]
> since we cant use serverSideProps together with initialProps. we can make use of React FunctionComponent.defaultProps
> here is the workaround that works for me, and it still works with initialProps ([#615 (comment)](https://github.com/isaachinman/next-i18next/issues/615#issuecomment-575578375))
> 
> 1. particular page
> 
> ```
> Page.defaultProps = {
>       i18nNamespaces: ['home-page']
> }
> ```
> 
> 1. main app / _app.tsx
> 
> ```
> MyApp.getInitialProps = async (appContext: AppContextType<Router>) => {
>     const appProps = await App.getInitialProps(appContext)
>     const defaultProps = appContext.Component.defaultProps
>     return {
>         ...appProps,
>         pageProps: {
>             namespacesRequired: [...(appProps.pageProps.namespacesRequired || []), ...(defaultProps?.i18nNamespaces || [])]
>         }
>     }
> }
> ```

this works. thanks!

**Comment by rparree** [comment]
Is it correct that the `defaultProps` must be set on the result of the HOC:

This works for me:

```ts
const page = withTranslation('toto')(TotoComponent);

page.defaultProps = {
    i18nNamespaces: ['toto','bar']
}

export default page;
```

This does not:

```ts
TotoComponent.defaultProps = {
    i18nNamespaces: ['toto','bar']
}

export default withTranslation('toto')(TotoComponent);
```

**Comment by rparree** [comment]
I'm using this version, so that it picks up recursive namespaces from WrappedComponents when using HOC

```ts
MyApp.getInitialProps = async (appContext: AppContextType<Router>) => {

    const recursiveNamespaces = (component: NextComponentType & { WrappedComponent?: NextComponentType }, acc: string[] = []): string[] => {

        const ns = component.defaultProps?.i18nNamespaces;
        const a = (typeof ns === "string") ? [...acc, ns] : [...acc, ...(ns || [])];
        let wrappedComponent = component.WrappedComponent;
        if (wrappedComponent) {
            return recursiveNamespaces(wrappedComponent, a)
        } else
            return a
    }

    const appProps = await App.getInitialProps(appContext)
    const namespaces: any = recursiveNamespaces(appContext.Component)
    return {
        ...appProps,
        pageProps: {
            namespacesRequired: [
                ...(appProps.pageProps.namespacesRequired || []),
                ...namespaces]
        }
    }
}

```

**Comment by isaachinman** [comment]
@rparree I think you're commenting on the wrong issue. This issue is about `getServerSideProps`.

**Comment by rparree** [comment]
@isaachinman I was referring to the comment with the workaround submitted by @icrona above  (https://github.com/isaachinman/next-i18next/issues/652#issuecomment-644618517)

**Comment by isaachinman** [comment]
Sorry, what exactly is that workaround achieving?

**Comment by rparree** [comment]
With the solution in their comment you can use `defaultProps` to set the namespaces when using a component that uses `getServerSideProps`. My code looks for `defaultProps` recursively when using a HOC.

**Comment by isaachinman** [comment]
As far as I can tell, that doesn't achieve anything except slightly modifying the API. It won't enable SSG.

**Comment by sebastienlabine** [comment]
> since we cant use serverSideProps together with initialProps. we can make use of React FunctionComponent.defaultProps
> here is the workaround that works for me, and it still works with initialProps ([#615 (comment)](https://github.com/isaachinman/next-i18next/issues/615#issuecomment-575578375))
> 
> 1. particular page
> 
> ```
> Page.defaultProps = {
>       i18nNamespaces: ['home-page']
> }
> ```
> 
> 1. main app / _app.tsx
> 
> ```
> MyApp.getInitialProps = async (appContext: AppContextType<Router>) => {
>     const appProps = await App.getInitialProps(appContext)
>     const defaultProps = appContext.Component.defaultProps
>     return {
>         ...appProps,
>         pageProps: {
>             namespacesRequired: [...(appProps.pageProps.namespacesRequired || []), ...(defaultProps?.i18nNamespaces || [])]
>         }
>     }
> }
> ```

how would that work for typescript?

**Comment by rparree** [comment]
@isaachinman It does the trick for me. I am able to use a component like this (and it will be SSR)

```ts
export const getServerSideProps = async function (ctx: NextPageContext) {
  …
 return {
    props: {
      totoProp : ""
   }
 } 
}

TotoComponent.defaultProps = {
    i18nNamespaces: ['toto','bar']
}

export default withTranslation('toto')(TotoComponent);
```

**Comment by mattcrn** [comment]
> 
> 
> since we cant use serverSideProps together with initialProps. we can make use of React FunctionComponent.defaultProps
> here is the workaround that works for me, and it still works with initialProps ([#615 (comment)](https://github.com/isaachinman/next-i18next/issues/615#issuecomment-575578375))
> 
>     1. particular page
> 
> 
> ```
> Page.defaultProps = {
>       i18nNamespaces: ['home-page']
> }
> ```
> 
>     1. main app / _app.tsx
> 
> 
> ```
> MyApp.getInitialProps = async (appContext: AppContextType<Router>) => {
>     const appProps = await App.getInitialProps(appContext)
>     const defaultProps = appContext.Component.defaultProps
>     return {
>         ...appProps,
>         pageProps: {
>             namespacesRequired: [...(appProps.pageProps.namespacesRequired || []), ...(defaultProps?.i18nNamespaces || [])]
>         }
>     }
> }
> ```

This approach breaks SSR for me.

**Comment by mattcrn** [comment]
To get all namespaces loaded in case none are defined on the page, you would have to adapt icronas solution like this:
```js
NextApp.getInitialProps = async (appContext) => {
  const appProps = await App.getInitialProps(appContext);
  const { defaultProps } = appContext.Component;
  let pageProps = {}
  if(appProps.pageProps.namespacesRequired || defaultProps?.i18nNamespaces) {
    pageProps = {
      namespacesRequired: [...(appProps.pageProps.namespacesRequired || []), ...(defaultProps?.i18nNamespaces || [])],
    }
  }
  return {
    ...appProps,
    pageProps
  };
};
```

**Comment by isaachinman** [comment]
@mattcrn The `next-i18next` package will already load in all namespaces if you don't define any on the page. Let's keep this issue strictly about the support of `getServerSideProps`.

**Comment by mattcrn** [comment]
@isaachinman yes, but using icrona's solution will break this functionality since it will return an empty array for `namespacesRequired` and the package is simpy checking if `namespacesRequired`  is defined for loading all namespaces.
So, `namespacesRequired: []` means no namespaces are being loaded.

**Comment by mattcrn** [comment]
> 
> 
> @pettomartino Do you mean contributing to `next-i18next`? If we want to support `getServerSideProps`, we'll need to update the `appWithTranslation` HOC to also look for that data fetching method in addition to `getInitialProps`.

As far as I understand it, getServerSideProps is not part of the page component, so we would have to create a higher order function to wrap the pages getServerSideProps. Or is there a better way?

**Comment by ckeeney** [comment]
I think when this feature lands, it would be ideal to support both `getServerSideProps` and `getStaticProps`.  For most users of this library, local translations and staticProps will be fine, but the serverSideProps may be useful in projects that are using https://locize.com/ or something similar, where translations are managed outside of the project.

**Comment by Divlo** [comment]
Any update since then ?
Looking forward to have this thing implemented. :+1:

**Comment by isaachinman** [comment]
@Divlo Feel free to submit a PR.

**Comment by maxgfr** [comment]
+1

**Comment by SalahAdDin** [comment]
> since we cant use serverSideProps together with initialProps. we can make use of React FunctionComponent.defaultProps
> here is the workaround that works for me, and it still works with initialProps ([#615 (comment)](https://github.com/isaachinman/next-i18next/issues/615#issuecomment-575578375))
> 
> 1. particular page
> 
> ```
> Page.defaultProps = {
>       i18nNamespaces: ['home-page']
> }
> ```
> 
> 1. main app / _app.tsx
> 
> ```
> MyApp.getInitialProps = async (appContext: AppContextType<Router>) => {
>     const appProps = await App.getInitialProps(appContext)
>     const defaultProps = appContext.Component.defaultProps
>     return {
>         ...appProps,
>         pageProps: {
>             namespacesRequired: [...(appProps.pageProps.namespacesRequired || []), ...(defaultProps?.i18nNamespaces || [])]
>         }
>     }
> }
> ```

Well, i thought it would work good, but it works with delay:
![Screen Capture_select-area_20201201232330](https://user-images.githubusercontent.com/5034215/100793054-a5cfda80-342c-11eb-9493-1c493057b88d.gif)

At the first loading, it does not load the `skill-test` namespaces, i need to reload it to get that namespace active and the proper localization strings.

Inspecting it for the first time we get this:
![Screen Capture_select-area_20201201232220](https://user-images.githubusercontent.com/5034215/100793222-dc0d5a00-342c-11eb-8217-fdc183a570c5.jpg)

After reloading, we can find the required namespace:
![image](https://user-images.githubusercontent.com/5034215/100793400-17a82400-342d-11eb-999f-9289e1f6ea9a.png)

Both namespaces are in the option, but just the first one, defined on `_app.js` by `getInitialProps`, is loaded.

Why does this happen?

**Comment by isaachinman** [comment]
@SalahAdDin This will be solved by #869.

**Comment by isaachinman** [comment]
Fixed in `next-i18next@8.0.0`.

**Comment by belgattitude** [comment]
Thanks a lot for this @isaachinman

**Comment by SalahAdDin** [comment]
@isaachinman Does the new release mean we must deleted the whole workaround for this matter and use the new hook? It is not so much clear for me.

**Comment by isaachinman** [comment]
@SalahAdDin Correct, there is no need for a workaround in v8.

**Comment by syamjayaraj** [comment]
Import the **serverSideTranslations**

`import { serverSideTranslations } from "next-i18next/serverSideTranslations";
`

Now from the **getServerSideProps**, pass the .`.(await serverSideTranslations(locale, ["common"])),` with the props.

```
export const getServerSideProps: GetStaticProps = async ({
  locale,
  locales,
  query
}: any) => {
  return {
    props: {
      ...(await serverSideTranslations(locale, ["common"])),
    }
  }
};
```

Now add your language string inside
**/public/locales/en/common.json**

For example

```
{
  "home": {
    "Test": "Test"
  }
}
```

You can add more language strings inside the locales directory.

Restart your Next.js app. It will work.
